File: apps/ios/Sources/SwabUI/Carte/CarteViewModel.swift
/// MAP-05 — carte state: loads from the on-device vault ONLY. This file
/// deliberately imports nothing beyond `Observation`/`SwabCore` — no
/// networking import exists here, by construction. `CarteOfflineInvariantTests`
/// (SwabCoreTests) scans this file's source text as a structural guard
/// against a future regression, mirroring the RN reference's own MAP-05
/// test that scans for API imports. (Deliberately not naming the specific
/// banned symbols in this comment — the scanner matches on them literally.)
import Observation
import SwabCore

@MainActor
@Observable
public final class CarteViewModel {
    /// SUG-IOS-004 / VLT-01 / MAP-06: distinguishes "no data yet"/"genuinely
    /// no contacts" from "data exists but couldn't be read" — the two must
    /// never render the same calm-empty copy (that was silent data loss).
    public enum LoadState: Equatable {
        case loading, loaded, unreadable
    }

    public private(set) var contacts: [VaultContact] = []
    public private(set) var loadState: LoadState = .loading
    public var listMode = false
    public var legendOpen = false
    public var selected: VaultContact?

    private let vault: Vault
    private let reporter: ErrorReporter

    public init(vault: Vault, reporter: ErrorReporter = NoopErrorReporter()) {
        self.vault = vault
        self.reporter = reporter
    }

    public var unplaced: [VaultContact] {
        contacts.filter { $0.ring == nil }
    }

    public var placed: [VaultContact] {
        contacts.filter { $0.ring != nil }
    }

    /// MAP-05: zero network, offline-first. Called on appear and whenever
    /// the app returns to foreground, so an FS-03 re-tag is reflected with
    /// an animated move rather than requiring a relaunch.
    public func refresh() async {
        do {
            contacts = try await vault.getContacts()
            loadState = .loaded
        } catch {
            reporter.report(
                ReportedError(domain: "carte.vault", operation: "refresh", errorDescription: VaultError.reportCode(for: error))
            )
            contacts = []
            loadState = .unreadable
        }
    }

    public func select(_ contact: VaultContact) {
        selected = contact
    }

    public func closeSheet() {
        selected = nil
    }

    public func toggleLegend() {
        legendOpen.toggle()
    }

    /// FS-03 seam: constructs the fiche's view model without exposing this
    /// view model's private `vault` reference — `CarteView` calls this when
    /// « Ouvrir la fiche » is tapped instead of holding its own `Vault`.
    public func makeFicheViewModel(for contact: VaultContact) -> FicheViewModel {
        FicheViewModel(vault: vault, contact: contact, reporter: reporter)
    }
}

File: apps/android/app/src/main/kotlin/com/swab/android/carte/CarteViewModel.kt
package com.swab.android.carte

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.swab.android.vault.Vault
import com.swab.android.vault.VaultContact
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * FS-02 « Ma carte » — MAP-01/05: state for the map/list screen, sourced
 * ENTIRELY from the on-device [Vault]. Port of apps/mobile/app/(main)/carte.tsx.
 *
 * MAP-05 (offline by construction): this file must never import
 * com.swab.android.network — enforced structurally by
 * CarteOfflineStructuralTest, which scans this package's source for
 * forbidden imports the same way the RN reference's carte-offline.map05
 * test scans the map and (main) source trees.
 */
class CarteViewModel(private val vault: Vault) : ViewModel() {

    private val _contacts = MutableStateFlow<List<VaultContact>>(emptyList())
    val contacts: StateFlow<List<VaultContact>> = _contacts.asStateFlow()

    private val _listMode = MutableStateFlow(false)
    val listMode: StateFlow<Boolean> = _listMode.asStateFlow()

    private val _legendOpen = MutableStateFlow(false)
    val legendOpen: StateFlow<Boolean> = _legendOpen.asStateFlow()

    private val _selected = MutableStateFlow<VaultContact?>(null)
    val selected: StateFlow<VaultContact?> = _selected.asStateFlow()

    init {
        refresh()
    }

    /**
     * Reload from the vault. Called on init, and should be called again
     * whenever the carte regains focus/resumes (RN's useFocusEffect
     * equivalent) so an FS-03 re-tag shows the contact on its new ring with
     * an animated move on return (MAP-04 acceptance criterion) — the FS-03
     * screen itself doesn't exist yet, but this refresh seam is ready.
     */
    fun refresh() {
        viewModelScope.launch { _contacts.value = vault.getContacts() }
    }

    fun setListMode(value: Boolean) {
        _listMode.value = value
    }

    fun toggleLegend() {
        _legendOpen.value = !_legendOpen.value
    }

    fun selectContact(contact: VaultContact) {
        _selected.value = contact
    }

    fun clearSelection() {
        _selected.value = null
    }
}
